lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -892,6 +902,9 @@ public int resolveUnion(Schema union, Object datum) {
   protected String getSchemaName(Object datum) {
     if (datum == null || datum == JsonProperties.NULL_VALUE)
       return Type.NULL.getName();
+    String primativeType = PRIMATIVE_DATUM_TYPES.get(datum.getClass());
+    if (primativeType != null)
+      return primativeType;
```

Comment:
Shouldn't we remove the later checks for `isInteger(datum)`, `isLong(datum)`, etc. ?

Reply:
Hello,

It was my desire to remove these methods, but I feared I would be breaking backwards compatibility (even more than this already does).  It may still be possible for developers to override these methods. For example:

```java
    if (isInteger(datum))
      return Type.INT.getName();
```

A developer could allow for the primitive case to be handled by the cache, but also extend the functionality to allow for other logical types to, in this case, return an INT type by overriding `isInteger`.  I'm happy to take away that ability if it offers no real-world use cases, but I didn't want to be so heavy-handed with it.

Reply:
I am not sure whether I was clear enough.
https://github.com/apache/avro/blob/cc0eb0bee35681f33a0b5324b106ef744be86d1f/lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java#L905-L907 is the new `if` branch.
If I understand it correctly if the datum is any of the "primitive" types then the following `if` branches won't ever be reached: https://github.com/apache/avro/blob/cc0eb0bee35681f33a0b5324b106ef744be86d1f/lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java#L918-L931
So, even if a user overrides `isInteger()` it won't be called at all, because of https://github.com/apache/avro/blob/cc0eb0bee35681f33a0b5324b106ef744be86d1f/lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java#L905-L907

Reply:
In addition, `PRIMATIVE_DATUM_TYPES` handles primitives' wrapper classes, like `Integer.class`. Does it need to handle `Integer.TYPE` too ?

Reply:
I guess the subsequent code wouldn't be **entirely** dead, since somebody might have an implementation that turns their special `MyInteger` class into a `Type.INT`.  Hypothetically, this change would make it impossible for a subclass to specify that all `java.lang.Integer` should be represented by a `Type.LONG`, I guess! 

While I doubt anybody is actually doing that, why don't we play it with a bit of security and do: 

```
String primativeType = getPrimitiveTypeCache().get(datum.getClass());
```

where `getPrimitiveTypeCache()` is protected and returns `PRIMATIVE_DATUMS_TYPES` by default?  That adds one method call but would give a subclass an "escape route" if they do have a specialised behaviour.  Alternatively, a protected data member that is initialised to the static hash instance perhaps?